p2p/src/pubsub.rs

```diff
@@ -149,90 +111,142 @@ where
             .await
             .map_err(|_| P2pError::SubsystemFailure)?;
 
+        self.pubsub_handle.subscribe(&self.topics).await?;
+        Ok(rx)
+    }
+
+    /// Process block announcement from the network
+    async fn process_block_announcement(
+        &mut self,
+        peer_id: T::PeerId,
+        message_id: T::MessageId,
+        block: Block,
+    ) -> crate::Result<()> {
+        let result = self
+            .chainstate_handle
+            .call_mut(move |this| this.process_block(block, chainstate::BlockSource::Peer))
```

Comment:
Please call preliminary_block_check() before calling process_block() (separately, not in the same lambda). The reason is that the precheck is immutable, which means it can run in parallel and not block the queue of chainstate.

Reply:
I think this would require making a clone of each block, regardless of whether there would be contention with calls to chainstate. Seems awfully wasteful

Reply:
That's fair. Maybe we should use Arc. Another solution is to ping-pong Block back and forth. So it can be returned in the return value of preliminary_block_check().

Reply:
I like the latter idea better, I'll use that.